### PR TITLE
Convert route params array into object

### PR DESCRIPTION
**Pull request by eileencodes**
This PR converts the route params array into an object and moves the
error generation code into its own object as well. This is a refactoring
of internal code to make it easier to change and simplier for internals
to interface with. We have two new objects:

1) `RouteWithParams`

Previously `#generate` returned an array of parameterized parts for a
route. We have now turned this into an object with a `path` and a
`params methods` to be able to access the parts we need.

2) `MissingRoute`

`#generate` was also responsible for constructing the message for the
error. We've moved this code into the new `MissingRoute` object so it
does that work instead.

This change makes these methods reusable throughout the code and easier
for future refactoring we plan to do.

Co-authored-by: Aaron Patterson <aaron.patterson@gmail.com>

---

Note We haven't included a changelog because all of these APIs are private. Both files in lib with changes have `stopdoc` at the top.

### CODE REVIEW

**Comment by jonathanhefner** [comment]
> This change makes these methods reusable throughout the code and easier
> for future refactoring we plan to do.

Cool!  Are you eyeing #32887?  I was digging into it a while back, and I recall thinking a proper solution would require a significant refactor, although I don't recall the specifics. :pensive:

File: actionpack/lib/action_dispatch/routing/route_set.rb [hunk 1]
```diff
@@ -735,8 +735,8 @@ def normalize_controller!
 
         # Generates a path from routes, returns [path, params].
         # If no route is generated the formatter will raise ActionController::UrlGenerationError
```
Comment (jonathanhefner) [inline actionpack/lib/action_dispatch/routing/route_set.rb:737]:
```suggestion
        # Generates a path from routes, returns a RouteWithParams or MissingRoute.
```

**Comment by tenderlove** [comment]
> Cool! Are you eyeing #32887? I was digging into it a while back, and I recall thinking a proper solution would require a significant refactor, although I don't recall the specifics.

No.  We're trying to refactor routing internals so that we can modify and or remove Journey.  One goal is to get the routes to raise an exception closer to user-land code, another goal is convert tests to be more "integration" style (for example `generate_extras` and `extra_keys` are never called in an actual app, so calling them in test doesn't make sense), and the final goal is speed.

File: actionpack/lib/action_dispatch/routing/route_set.rb (hunk 1 above)
Comment (eileencodes) [reply actionpack/lib/action_dispatch/routing/route_set.rb]: Fixed!

**Comment by Edouard-chin** [comment]
Hello friends 👋 , I think this PR created a regression , the reproduction script is here https://gist.github.com/Edouard-chin/8425a6e57a3e0746e5431cd1543ff8ce 

TL;DR If you pass a nested param in the request where the top key is one of the `RESERVED_OPTIONS` (domain, host ...) , i.e `params: { domain: { foo: 'bar ' } }`, the params in the controller won't be nested like they used to. 
```ruby
def my_action
  puts params['domain']['foo']
   # before
   # => bar
   #
   # after
   # TypeError: no implicit conversion of Symbol into Integer
   puts params['domain'] # => fooo=bar
end
